server/src/main/java/org/opensearch/index/codec/CodecService.java

```diff
@@ -71,10 +75,11 @@ public CodecService(@Nullable MapperService mapperService, Logger logger) {
             codecs.put(ZSTD_CODEC, new ZstdCodec());
             codecs.put(ZSTD_NO_DICT_CODEC, new ZstdNoDictCodec());
         } else {
+            int compressionLevel = mapperService.getIndexSettings().getValue(INDEX_CODEC_COMPRESSION_LEVEL_SETTING);
```

Comment:
Should we have the constructor that accepts `indexSettings`? `public CodecService(@Nullable MapperService mapperService , IndexSettings indexSettings, Logger logger)`, the `mapperService` could be `null` but `indexSettings` should be always available I think.

Reply:
> @reta I thought about it, but I saw that MapperService is always initialized over [here](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/index/IndexService.java#L224), hence thought it may not be required to have another ctor.

It is a little bit more nuanced, the `CodecService` may come from the plugins (you could see that by inspecting the `EnginePlugin`), there the `MapperService` may not be available (hence this is why it is `@Nullbable`) but `IndexSettings` are. By splitting these apart, we could also provide compression level to the codecs even if the mapper service is not provided. (the `if` branches)

Reply:
yeah makes sense. I'll refactor this

Reply:
@reta, wanted to point out that once I refactor the CodecService constructor, the diff will be quite big considering I will have to refactor ```CorrelationCodecService``` in events-correlation-engine plugin and around 15-20 test classes.

Reply:
updated the PR with the changes